sympy/matrices/random.py

```diff
@@ -0,0 +1,2388 @@
+          'diagonal_normal', 'jordan_normal', 'isometry_normal', \
+          'triangular', 'invertible', 'singular', \
+          'idempotent', 'nilpotent', 'diagonalizable', 'triangularizable', \
+          'orthogonal', 'unitary', 'normal', \
+          'symmetric', 'hermite', 'square', \
+          'regular_to_singular', 'complex_to_real'
+
+
+# === random number generator functions ===
+
+
+def _sample(scalars, k=None):
+    _k = k or 1
+    if hasattr(scalars, 'sample'):
+        smpl = scalars.sample(_k)
+    else:
+        smpl = sample(scalars, _k)
+    return smpl if k else smpl[0]
+
+
+# === helper ===
+
+
+_EPS = 2 * abs(abs(complex(1, 1) / abs(complex(1, 1))) - 1)
+"""numerical tolerance for float and complex types"""
+
+
+def _is_real(z):
+    """checks number to be real number, incl. complex without imaginary part"""
+    if isinstance(z, (int, float)):
+        return True
+    if isinstance(z, complex):
+        return not z.imag
+    z = sympify(z)
+    return not bool(z.is_complex) or bool(z.is_real)
+
+
+# === default sets ===
+
+
+_elementary_scalars = -1, 1
+_elementary_units = _elementary_scalars
+_rotation_scalars = (sqrt(2) / 2, sqrt(2) / 2), (0, -1),
+_unitary_scalars = tuple(c * 1 + s * I for c, s in _rotation_scalars)
+
+
+# === fundamental constructor ===
+
+
+def super_elementary(dim,
+                     index=None,
+                     value=None,
+                     *scalars):
+    r"""super elementary matrix n x n, i.e. identiy with a 2 x 2 block
+
+    Explanation
+    ===========
+    The super elementary matrix $A$ is a generalization of elementary matrices
+    as well as of a rotation matrices that rotate only a single plane.
+
+    In two dimensions any invertible matrix
+
+    .. math::
+
+       A = \left[\begin{array}{cc}a & b \\ -c & d\end{array}\right]
+
+    is super elementary, i.e. $ad+bc \neq 0$. In higher dimensions,
+    any matrix $A$ looking like itdentiy matrix but with entries
+
+    .. math::
+
+        A[i,i] = a, \quad A[i,j] = b, \quad A[j,i] = -c, \quad A[j,j] = d
+
+    and $ad + bc \neq 0$.
+
+    This includes elementary matrices of matrix operation
+    for Gauss elimination. So multiplication with $A$ gives for
+
+    * $a=0=d$ and $b=1=-c$ a *transposition*, i.e. swapping rows $i$ and $j$
+    * $a=1$, $d=\lambda$ and $b=0=-c$ this matrix describes scaling the
+      row $j$ by $\lambda$
+    * $a=1=d$, $b=\lambda$, $-c=0$ adding the $\lambda$ multiple
+      of the row $i$ to row $j$.
+
+    Moreover, a simple rotation by $\phi$ in $(i,j)$ plane is super elementary,
+    given by $a=\cos \phi = d$ and $b = \sin \phi = c$. Hence,
+
+    .. math::
+
+       \left[\begin{array}{cc}a & b \\ -c & d\end{array}\right]
+       =
+       \left[\begin{array}{cc}
+        \cos \phi & \sin \phi \\ -\sin \phi & \cos \phi \end{array}\right]
+
+    In this module the super elementary matrix serves as the base class to
+    create futher matrices of given type.
```

Comment:
Perhaps "base case" because this isn't really a class.

Reply:
Best would be to remove this function. It is kind of legacy as it is called only from `elementary`, `transposition` and `rotation` and each has it own codepath within `super_elementary`. It is more clear to create those matrices explicitly within each function.